beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+	}
+
+	defer func() {
+		closeErr := file.Close()
+
+		// Overwrite the existing error only if it is nil, since the close error is less important.
+		if closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+
+	// Encode the SSZ encoded data column sidecar size.
+	var encodedSszEncodedDataColumnSidecarSize [encodedSszEncodedDataColumnSidecarSizeSize]byte
+	binary.BigEndian.PutUint32(encodedSszEncodedDataColumnSidecarSize[:], uint32(sszEncodedDataColumnSidecarRefSize))
+
+	// Concatenate the version, the data column sidecar size, the data column indices and the SSZ encoded data column sidecar.
+	countToWrite := headerSize + len(sszEncodedDataColumnSidecars)
+	bytes := make([]byte, 0, countToWrite)
+	bytes = append(bytes, byte(version))
+	bytes = append(bytes, encodedSszEncodedDataColumnSidecarSize[:]...)
+	bytes = append(bytes, indices[:]...)
+	bytes = append(bytes, sszEncodedDataColumnSidecars...)
+
+	countWritten, err := file.Write(bytes)
+	if err != nil {
+		return errors.Wrap(err, "write")
+	}
+	if countWritten != countToWrite {
+		return ErrWrongBytesWritten
+	}
+
+	return nil
+}
+
+// metadata runs file sanity checks and retrieves metadata of the file.
+// The file descriptor is left at the beginning of the first SSZ encoded data column sidecar.
+func (dcs *DataColumnStorage) metadata(file afero.File) (*metadata, error) {
+	// Seek to the beginning of the file.
+	_, err := file.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, errors.Wrap(err, "seek")
+	}
+
+	// Read the encoded file version.
+	var encodedFileVersion [versionSize]byte
+	countRead, err := file.Read(encodedFileVersion[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read file version")
+	}
+
+	if countRead != versionSize {
+		return nil, ErrWrongBytesVersionRead
+	}
+
+	// Convert the version to an int.
+	fileVersion := int(encodedFileVersion[0])
+
+	// Check if the version is the expected one.
+	if fileVersion != version {
+		return nil, ErrWrongVersion
+	}
+
+	// Read the SSZ encoded data column sidecar size.
+	var encodedSszEncodedDataColumnSidecarSize [encodedSszEncodedDataColumnSidecarSizeSize]byte
+	countRead, err = file.Read(encodedSszEncodedDataColumnSidecarSize[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read SSZ encoded data column sidecar size")
+	}
+	if countRead != encodedSszEncodedDataColumnSidecarSizeSize {
+		return nil, ErrWrongBytesDataColumnSidecarSizeRead
+	}
+
+	// Convert the SSZ encoded data column sidecar size to an int.
+	sszEncodedDataColumnSidecarSize := binary.BigEndian.Uint32(encodedSszEncodedDataColumnSidecarSize[:])
+
+	// Read the data column indices.
+	var indices [mandatoryNumberOfColumns]byte
+	countRead, err = file.Read(indices[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read data column indices")
+	}
+	if countRead != mandatoryNumberOfColumns {
+		return nil, ErrWrongBytesIndicesRead
+	}
+
+	// Retrieve the statistics of the file.
+	fileStat, err := file.Stat()
```

Comment:
See below comment on `Size` - `Stat` can be expensive, like opening another file.

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/commit/22b3327a97afc4b1a171f569ce24520520b66a26.